File: blender_roblox_sync/roblox_interface/roblox_server/format_types/unitvec.py
import struct, math, mathutils

YAW = 32767 / math.pi
PITCH = 32767 / (0.5 * math.pi)
INV_YAW = 1 / YAW
INV_PITCH = 1 / PITCH

format = struct.Struct("<2h")
format_size = format.size
unpack_from = format.unpack_from
pack = format.pack

def read(self, format_data, **masks):
	normal_yaw, normal_pitch = unpack_from(*self.read_buffer(format_size))
	normal_yaw *= INV_YAW
	normal_pitch *= INV_PITCH
	horizontal = math.cos(normal_pitch)
	return (mathutils.Vector((math.sin(normal_yaw) * horizontal, math.cos(normal_yaw) * horizontal, math.sin(normal_pitch))),)

def write(self, args, args_count, format_data, **masks):
	vec_value = args[args_count]
	normal_yaw = math.atan2(vec_value.x, vec_value.y)
	normal_pitch = math.asin(vec_value.z)
	self.write_buffer(pack(
		round(normal_yaw * YAW),
		round(normal_pitch * PITCH)
	), format_size)
	return args_count + 1

File: roblox_plugin/src/Server/formatTypes/unitvec.ts
//!native
import type * as ProcessFormats from "../processFormats";
import SendMessagesThread from "../sendMessages";
import ReceiveMessagesThread from "../receiveMessages";

const YAW = 32767 / math.pi;
const INV_YAW = 1 / YAW;
const PITCH = 32767 / (0.5 * math.pi);
const INV_PITCH = 1 / PITCH;

const DATA_SIZE = 4;
export function read(
	receiveThread: ReceiveMessagesThread,
	args: defined[],
	formatData: ProcessFormats.FormatData,
	masks: Map<string, boolean>,
) {
	const [readBuffer, readOffset] = ReceiveMessagesThread.readBuffer(receiveThread, DATA_SIZE);
	const normalYaw = buffer.readi16(readBuffer, readOffset) * INV_YAW;
	const normalPitch = buffer.readi16(readBuffer, readOffset + 2) * INV_PITCH;
	const horizontal = math.cos(normalPitch);
	args.push(new Vector3(math.cos(normalYaw) * horizontal, math.sin(normalPitch), math.sin(normalYaw) * horizontal));
}
export function write(
	sendThread: SendMessagesThread,
	args: defined[],
	argsCount: number,
	formatData: ProcessFormats.FormatData,
	masks: Map<string, boolean>,
): number {
	const value = args[argsCount] as Vector3;
	const velMag = value.Magnitude;
	const unitVel = velMag !== 0 ? value.div(velMag) : Vector3.xAxis;
	const normalYaw = math.atan2(unitVel.Z, unitVel.X);
	const normalPitch = math.asin(unitVel.Y);

	const data = buffer.create(DATA_SIZE);
	buffer.writei16(data, 0, math.round(normalYaw * YAW));
	buffer.writei16(data, 2, math.round(normalPitch * PITCH));
	SendMessagesThread.writeBuffer(sendThread, data, DATA_SIZE);
	return argsCount + 1;
}
